akka-projection-slick/src/main/scala/akka/projection/slick/SlickProjection.scala

```diff
@@ -23,19 +24,30 @@ object SlickProjection {
   def exactlyOnce[Offset, Envelope, P <: JdbcProfile: ClassTag](
       projectionId: ProjectionId,
       sourceProvider: SourceProvider[Offset, Envelope],
-      databaseConfig: DatabaseConfig[P])(eventHandler: Envelope => DBIO[Done]): Projection[Envelope] =
+      databaseConfig: DatabaseConfig[P],
+      eventHandler: SlickEventHandler[Envelope]): Projection[Envelope] =
     new SlickProjectionImpl(projectionId, sourceProvider, databaseConfig, SlickProjectionImpl.ExactlyOnce, eventHandler)
 
   def atLeastOnce[Offset, Envelope, P <: JdbcProfile: ClassTag](
       projectionId: ProjectionId,
       sourceProvider: SourceProvider[Offset, Envelope],
       databaseConfig: DatabaseConfig[P],
       saveOffsetAfterEnvelopes: Int,
-      saveOffsetAfterDuration: FiniteDuration)(eventHandler: Envelope => DBIO[Done]): Projection[Envelope] =
+      saveOffsetAfterDuration: FiniteDuration,
+      eventHandler: SlickEventHandler[Envelope]): Projection[Envelope] =
     new SlickProjectionImpl(
       projectionId,
       sourceProvider,
       databaseConfig,
       SlickProjectionImpl.AtLeastOnce(saveOffsetAfterEnvelopes, saveOffsetAfterDuration),
       eventHandler)
+
+}
+
+trait SlickEventHandler[Envelope] {
+  def handleEvent(envelope: Envelope): DBIO[Done]
+  def onFailure(envelope: Envelope, throwable: Throwable): Future[Done] = {
```

Comment:
I was planning to return DBIO[Done] in this function and do the recovery inside Slick. 

Unfortunately, Slick's `cleanUp` method seems to be broken. Or the documentation is misleading. It should behave like a `recoverWith`, but it will always fail the DBIO.

Reply:
Maybe here we take another approach and explore the idea of having RecoveryStrategies

* Skip
* Stop (default)
* Retry N times